nativelib/src/main/scala/scala/scalanative/native/time.scala

```diff
@@ -1,44 +1,77 @@
-  def ctime_s(buffer: Ptr[CChar], bufsz: CSize, time: Ptr[time_t]): CInt =
-    extern
-  def difftime(time_end: CLong, time_beg: CLong): CDouble      = extern
-  def gmtime(time: Ptr[time_t]): Ptr[tm]                       = extern
-  def gmtime_s(time: Ptr[time_t], result: Ptr[tm]): Ptr[tm]    = extern
-  def localtime(time: Ptr[time_t]): Ptr[tm]                    = extern
-  def localtime_s(time: Ptr[time_t], result: Ptr[tm]): Ptr[tm] = extern
-  def mktime(time: Ptr[tm]): time_t                            = extern
+  private final val timeBufferSize = 26 // taken from `man asctime`
+  @inline private def zeroAlloc[T](size: CSize)(implicit z: Zone): Ptr[T] = {
+    val buf = z.alloc(size)
+    string.memset(buf, 0, size)
+    buf.cast[Ptr[T]]
+  }
+
+  def asctime(time_ptr: Ptr[tm])(implicit z: Zone): CString =
+    asctime_r(time_ptr, zeroAlloc[CChar](timeBufferSize))
+  def asctime_r(time_ptr: Ptr[tm], buf: Ptr[CChar]): CString =
+    impl.asctime_r(time_ptr, buf)
+  def clock(): clock_t = impl.clock()
+  def ctime(time: Ptr[time_t])(implicit z: Zone): CString =
+    ctime_r(time, zeroAlloc[CChar](timeBufferSize))
+  def ctime_r(time: Ptr[time_t], buf: Ptr[CChar]): CString =
+    impl.ctime_r(time, buf)
+  def difftime(time_end: CLong, time_beg: CLong): CDouble =
+    impl.difftime(time_end, time_beg)
+  def gmtime(time: Ptr[time_t])(implicit z: Zone): Ptr[tm] =
+    gmtime_r(time, zeroAlloc[tm](sizeof[tm]))
+  def gmtime_r(time: Ptr[time_t], tm: Ptr[tm]): Ptr[tm] =
+    impl.gmtime_r(time, tm)
+  def localtime(time: Ptr[time_t])(implicit z: Zone): Ptr[tm] =
+    localtime_r(time, zeroAlloc[tm](sizeof[tm]))
+  def localtime_r(time: Ptr[time_t], result: Ptr[tm]): Ptr[tm] =
+    impl.localtime_r(time, result)
+  def mktime(time: Ptr[tm]): time_t = impl.mktime(time)
   def strftime(str: Ptr[CChar],
                count: CSize,
                format: CString,
-               time: Ptr[tm]): CSize                    = extern
-  def time(arg: Ptr[time_t]): time_t                    = extern
-  def timespec_get(ts: Ptr[timespec], base: CInt): CInt = extern
-  def tzset(): Unit                                     = extern
-  def wcsftime(str: CWideChar,
-               count: CSize,
-               format: Ptr[CWideChar],
-               time: Ptr[tm]): CSize = extern
-
-  @name("scalanative_daylight")
-  def daylight(): CInt = extern
-  @name("scalanative_timezone")
-  def timezone(): CLong = extern
-  @name("scalanative_tzname")
-  def tzname(): Ptr[CStruct2[CString, CString]] = extern
+               time: Ptr[tm]): CSize            = impl.strftime(str, count, format, time)
+  def time(arg: Ptr[time_t]): time_t            = impl.time(arg)
+  def tzset(): Unit                             = impl.tzset()
+  def daylight(): CInt                          = impl.daylight()
+  def timezone(): CLong                         = impl.timezone()
+  def tzname(): Ptr[CStruct2[CString, CString]] = impl.tzname()
+
+  @extern
+  private object impl {
```

Comment:
`object time` itself has to be extern with all of its methods beings extern. As mentioned above bindings should stay true to the original C API, not provide an additional layer of abstraction.

Reply:
Is there a technical reason for this or is it just the convention that scala native has adopted?

Reply:
We need raw bindings and we also need higher-level abstractions. By separating the two we make it clear which is which. Some people may require access to lowest-level APIs without any abstractions, while others might prefer to have ergonomics-oriented higher-level version of it. Both need to exist.

Reply:
That makes sense. Posix is what it is, for better and worse.